runtime/lua/vim/diagnostic.lua

```diff
@@ -1007,15 +1009,63 @@ local function next_diagnostic(search_forward, opts)
   -- Adjust row to be 0-indexed
   position[1] = position[1] - 1
 
+  -- Clamp column
+  position[2] = math.max(0, position[2])
+
   local wrap = if_nil(opts.wrap, true)
 
-  local diagnostics = get_diagnostics(bufnr, opts, true)
+  local diagnostics = {} ---@type vim.Diagnostic[]
+  ---ns -> extmark_id -> Diagnostic[]
+  local extmarks_index = {} ---@type table<integer, table<integer, vim.api.keyset.get_extmark_item>>
+  for namespace in pairs(diagnostic_cache[bufnr]) do
+    local ns = M.get_namespace(namespace)
+
+    local extmarks = api.nvim_buf_get_extmarks(
+      bufnr,
+      ns.user_data.location_ns,
+      { 0, 0 },
+      { -1, -1 },
+      {}
+    )
+
+    extmarks_index[namespace] = {}
+
+    for _, extmark in ipairs(extmarks) do
+      extmarks_index[namespace][extmark[1]] = extmark
+    end
+
+    local filtered_diagnostics =
+      filter_by_severity(opts.severity, diagnostic_cache[bufnr][namespace])
+
+    for _, diagnostic in ipairs(filtered_diagnostics) do
+      for _, extmark in ipairs(extmarks) do
+        if extmark[1] == diagnostic.extmark_id then
+          table.insert(diagnostics, diagnostic)
+        end
+      end
+    end
+  end
 
   if opts._highest then
     filter_highest(diagnostics)
   end
 
-  local line_diagnostics = diagnostic_lines(diagnostics)
+  local line_diagnostics = {} --- @type table<integer,vim.Diagnostic[]>
+  for _, diagnostic in ipairs(diagnostics) do
+    local namespace = diagnostic.namespace
+    assert(namespace, 'namespace must be set')
```

Comment:
You can combine the assignment and the `assert`:
```suggestion
    local namespace = diagnostic.namespace, 'namespace must be set')
```

Reply:
I guess there's a type in your suggestion?

Did you mean this f8de0867edf33656dc63e4bd5cd5fad6ccccbd00?

Reply:
Yes that. I shouldn't review PRs from my phone lol.